### PR TITLE
2098/fix partner application grid

### PR DESCRIPTION
**Pull request by YazeedLoonat**
## Issue Overview

This PR addresses [#issue](https://github.com/bloom-housing/bloom/issues/2098)

- [x] This change addresses the issue in full

## Description

This started as a rehash of https://github.com/bloom-housing/bloom/pull/2188 but somethings came up. Mainly there are big limitations to the nestjs-typeorm-pagination library that was causing issues where it was being used

I resolved this for partner's applications grid for a listing, and partner's user grid  

## How Can This Be Tested/Reviewed?
Please see the previous PR for instructions and comments, but a quick run is:

go to a listing on the partner's side and view the applications associated with that listing. The grid should display correct with records and pagination working properly

You should also test on the partner's side the user grid should now display data correctly which can be compared to what is displayed on the dev currently 

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [x] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

**Comment by emilyjablonski** [comment]
@YazeedLoonat Can you indicate the problems you were running into so I make sure to test those scenarios?

File: backend/core/src/applications/services/applications.service.ts [hunk 1]
```diff
@@ -66,14 +67,47 @@ export class ApplicationsService {
   async listPaginated(
     params: PaginatedApplicationListQueryParams
   ): Promise<Pagination<Application>> {
-    const qb = this._getQb(params)
-    const result = await paginate(qb, { limit: params.limit, page: params.page })
-    await Promise.all(
-      result.items.map(async (application) => {
-        await this.authorizeUserAction(this.req.user, application, authzActions.read)
+    if (!params.listingId) {
```
Comment (emilyjablonski) [inline backend/core/src/applications/services/applications.service.ts:70]: what's the paginated case where we're fetching applications without a listing id?

File: backend/core/test/user/user.e2e-spec.ts [hunk 2]
```diff
@@ -745,7 +745,7 @@ describe("Applications", () => {
       .get(`/user/list`)
       .set(...setAuthorization(adminAccessToken))
       .expect(200)
-    expect(allUsersListRes.body.meta.totalItems).toBe(totalUsersCount)
+    expect(allUsersListRes.body.meta.totalItems >= totalUsersCount).toBe(true)
```
Comment (emilyjablonski) [inline backend/core/test/user/user.e2e-spec.ts:748]: what's the case where the total items is greater than the total user count?

File: sites/partners/cypress/integration/application.spec.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+describe("Application Management Tests", () => {
```
Comment (emilyjablonski) [inline sites/partners/cypress/integration/application.spec.ts:1]: 🎈

**Comment by emilyjablonski** [comment]
LGTM locally! I am assuming this is unrelated to your changes, but sorting only sorted the current paginated view, not all items. Made issue #2283 but I'm curious if you think there is anything here prohibitive in fixing that in terms of how we are fetching the paginated items.

File: backend/core/src/applications/services/applications.service.ts (hunk 1 above)
Comment (YazeedLoonat) [reply backend/core/src/applications/services/applications.service.ts:70]: I'm honestly not sure, as it stands that endpoint doesn't require a listing id to be passed in
I can change it so we can drop that case but I wasn't sure if there was a reason for listing id to be not required

File: backend/core/test/user/user.e2e-spec.ts (hunk 2 above)
Comment (YazeedLoonat) [reply backend/core/test/user/user.e2e-spec.ts]: ah oops put that back the way it was my b

**Review by pbn4 — COMMENTED** [review]
We already discusses this a bit off github, @seanmalbert what do you think about rolling our own paginate solution?

**Comment by seanmalbert** [comment]
> We already discussed this a bit off github, @seanmalbert what do you think about rolling our own paginate solution?

@pbn4 ,
My initial reaction to your question is that I would prefer not to, but you have my curiosity. Can you lay out your proposal for this, so we can take it into consideration?

**Review by emilyjablonski — APPROVED** [review]
Sorting looks good to me thank you! Visually it's doing this not-ideal thing where the rows flicker in and out / move in order as we resort but I don't think we need to address that now 👍

**Comment by seanmalbert** [comment]
@pbn4 , adding a reminder for you to add your proposal here, so we can make a decision on how we want to proceed.

**Comment by pbn4** [comment]
@seanmalbert It seems that the issue with nestjs-typeorm-paginate has been outlined [here](https://github.com/nestjsx/nestjs-typeorm-paginate/issues/627) but it has not been fixed because it's not clear why commit 890ee164e15a1495782917b8443f0cd4cca0b7c0 which introduced this regression was introduced in the first place. 

How pagination is implemented in here is quite straightforward, it's just a method that does a merge of two queries and forward all repository / querybuilder options: 
```
async function paginateQueryBuilder<T, CustomMetaType = IPaginationMeta>(
  queryBuilder: SelectQueryBuilder<T>,
  options: IPaginationOptions<CustomMetaType>,
): Promise<Pagination<T, CustomMetaType>> {
  const [page, limit, route, paginationType, countQueries] =
    resolveOptions(options);

  const promises: [Promise<T[]>, Promise<number> | undefined] = [
    (PaginationTypeEnum.LIMIT_AND_OFFSET === paginationType
      ? queryBuilder.limit(limit).offset((page - 1) * limit)
      : queryBuilder.take(limit).skip((page - 1) * limit)
    ).getMany(),
    undefined,
  ];

  if (countQueries) {
    promises[1] = countQuery(queryBuilder);
  }

  const [items, total] = await Promise.all(promises);

  return createPaginationObject<T, CustomMetaType>({
    items,
    totalItems: total,
    currentPage: page,
    limit,
    route,
    metaTransformer: options.metaTransformer,
    routingLabels: options.routingLabels,
  });
}
```

So we could either recreate the same implementation in our source code and fix the count query as outlined in the issue or we could fork it and keep it as a package, which I believe is more work than the actual benefit we get from it.

**Comment by YazeedLoonat** [comment]
note to yazeed: peep dropping version of pagination library to see if we can use that

**Comment by YazeedLoonat** [comment]
@seanmalbert @pbn4 
From reading through the documentation I'm not sure that [commit](https://github.com/nestjsx/nestjs-typeorm-paginate/issues/627) is the main cause of the pagination issues:
the current version that we use of the nestjs-typeorm-paginate is: 2.2.1

which taking a peek at dev partner's user grid: https://dev-partners-bloom.netlify.app/users 

we see the same issue. That version is a year old, which is older than https://github.com/nestjsx/nestjs-typeorm-paginate/pull/610 which is the pr that the commit called out above got into master with

I even tried using versions like 2.6.2 which came out 4 months ago which is before the commit again, but more recent. 

That version gives us the same problem. So I'm of the opinion that the commit and the issue creator was actually wrong, and this problem is a bit more deeply tied to how typeorm functions

File: backend/core/src/applications/services/applications.service.ts (hunk 1 above)
Comment (seanmalbert) [reply backend/core/src/applications/services/applications.service.ts:70]: @YazeedLoonat , let's make listingId a required parameter.

**Review by seanmalbert — APPROVED** [review]
@YazeedLoonat,
As discussed I think we're good to merge this in, can you first please make listingId required for the application service's `listPaginated`. Make sure you add a test that tries calling it without the param. Once all tests are passing, I think you can go ahead and merge.

And make sure you run `generate:client` to update backend-swagger.

**Comment by seanmalbert** [comment]
@YazeedLoonat , we don't want to make listingId required after all, since @pbn4 brought up that we use the same endpoint to fetch a public user's applications. So we can merge this now.

**Pull request merged**